### PR TITLE
Match Torch speed for sum reduction on M1

### PR DESCRIPTION
**Pull request by nimlgen**
Match Torch speed for sum reduction on M1

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/helpers.py                        1               0
tinygrad/lazy.py                          24               3
tinygrad/ops.py                            3               3
tinygrad/runtime/lib.py                    8               1
tinygrad/runtime/ops_cuda.py               5               2
tinygrad/runtime/ops_gpu.py                3               2
tinygrad/runtime/ops_metal.py              3               2
------------------------------------------------------------
total                                     47              13
------------------------------------------------------------
lines added in the tinygrad folder: 34
```

**Comment by dc-dc-dc** [comment]
Can you post the speed results from `test/test_speed_v_torch.py`. Also a before/after of `test/models/test_real_world.py`

**Comment by nimlgen** [comment]
test_speed_v_torch:
```
add                                1x    1    0.44 ms (    0.00 GFLOPS     0.00 GB/s) in torch,    0.34 ms (    0.00 GFLOPS     0.00 GB/s) in tinygrad,    0.79x faster       0.00 MOPS     0.00 MB
add                             1024x 1024    0.56 ms (    1.88 GFLOPS    22.54 GB/s) in torch,    0.50 ms (    2.09 GFLOPS    25.11 GB/s) in tinygrad,    0.90x faster       1.05 MOPS    12.58 MB
add                             4096x 4096    1.73 ms (    9.72 GFLOPS   116.58 GB/s) in torch,    1.57 ms (   10.66 GFLOPS   127.87 GB/s) in tinygrad,    0.91x faster      16.78 MOPS   201.33 MB
add_constant                    4096x 4096    1.29 ms (   13.02 GFLOPS   104.20 GB/s) in torch,    1.18 ms (   14.23 GFLOPS   113.83 GB/s) in tinygrad,    0.92x faster      16.78 MOPS   134.22 MB
add_sq                          4096x 4096    3.26 ms (   15.46 GFLOPS    61.83 GB/s) in torch,    1.56 ms (   32.24 GFLOPS   128.97 GB/s) in tinygrad,    0.48x faster      50.33 MOPS   201.33 MB
array_packing                   2048x 2048    0.83 ms (    5.05 GFLOPS    40.36 GB/s) in torch,    0.62 ms (    6.75 GFLOPS    54.00 GB/s) in tinygrad,    0.75x faster       4.19 MOPS    33.55 MB
cat_0                            256x  256    0.46 ms (    0.29 GFLOPS     2.30 GB/s) in torch,    0.38 ms (    0.34 GFLOPS     2.74 GB/s) in tinygrad,    0.84x faster       0.13 MOPS     1.05 MB
cat_1                            256x  256    0.44 ms (    0.30 GFLOPS     2.38 GB/s) in torch,    0.40 ms (    0.33 GFLOPS     2.62 GB/s) in tinygrad,    0.91x faster       0.13 MOPS     1.05 MB
conv bs: 32 chans:  4 ->  32 k:3              0.75 ms (  100.64 GFLOPS     6.39 GB/s) in torch,    0.51 ms (  149.19 GFLOPS     9.47 GB/s) in tinygrad,    0.67x faster      75.50 MOPS     4.79 MB
conv bs: 32 chans: 16 ->  32 k:3              0.76 ms (  398.34 GFLOPS     8.68 GB/s) in torch,    0.87 ms (  345.53 GFLOPS     7.53 GB/s) in tinygrad,    1.15x slower     301.99 MOPS     6.58 MB
conv bs: 32 chans: 64 ->  32 k:3              1.30 ms (  929.11 GFLOPS    10.57 GB/s) in torch,    0.93 ms ( 1305.84 GFLOPS    14.85 GB/s) in tinygrad,    0.71x faster    1207.96 MOPS    13.74 MB
double_permute (64, 64, 64, 64)               2.92 ms (    5.75 GFLOPS    46.00 GB/s) in torch,    2.47 ms (    6.79 GFLOPS    54.35 GB/s) in tinygrad,    0.85x faster      16.78 MOPS   134.22 MB
exp                             2048x 2048    0.71 ms (   11.87 GFLOPS    47.46 GB/s) in torch,    0.63 ms (   13.29 GFLOPS    53.15 GB/s) in tinygrad,    0.89x faster       8.39 MOPS    33.55 MB
gemm                            1024x 1024    1.05 ms ( 2043.92 GFLOPS    11.98 GB/s) in torch,    0.94 ms ( 2273.77 GFLOPS    13.32 GB/s) in tinygrad,    0.90x faster    2147.48 MOPS    12.58 MB
gemm                             256x  256    0.46 ms (   72.27 GFLOPS     1.69 GB/s) in torch,    0.43 ms (   77.60 GFLOPS     1.82 GB/s) in tinygrad,    0.93x faster      33.55 MOPS     0.79 MB
gemm_unrolled                    512x  512    0.82 ms (  329.12 GFLOPS     3.86 GB/s) in torch,    0.74 ms (  364.31 GFLOPS     4.27 GB/s) in tinygrad,    0.90x faster     268.44 MOPS     3.15 MB
gemm_unrolled_permute_l          512x  512    0.59 ms (  453.12 GFLOPS     5.31 GB/s) in torch,    0.66 ms (  404.47 GFLOPS     4.74 GB/s) in tinygrad,    1.12x slower     268.44 MOPS     3.15 MB
gemm_unrolled_permute_lr         512x  512    0.55 ms (  488.14 GFLOPS     5.72 GB/s) in torch,    0.63 ms (  427.76 GFLOPS     5.01 GB/s) in tinygrad,    1.14x slower     268.44 MOPS     3.15 MB
gemm_unrolled_permute_r          512x  512    0.75 ms (  358.97 GFLOPS     4.21 GB/s) in torch,    0.45 ms (  596.30 GFLOPS     6.99 GB/s) in tinygrad,    0.60x faster     268.44 MOPS     3.15 MB
max                             4096x 4096    0.98 ms (   17.11 GFLOPS    68.72 GB/s) in torch,    0.88 ms (   19.14 GFLOPS    76.87 GB/s) in tinygrad,    0.89x faster      16.84 MOPS    67.63 MB
mul_sum                         4096x 4096    2.16 ms (   15.56 GFLOPS    62.37 GB/s) in torch,    1.26 ms (   26.78 GFLOPS   107.32 GB/s) in tinygrad,    0.58x faster      33.62 MOPS   134.74 MB
neg                             4096x 4096    1.29 ms (   13.04 GFLOPS   104.30 GB/s) in torch,    1.21 ms (   13.88 GFLOPS   111.05 GB/s) in tinygrad,    0.94x faster      16.78 MOPS   134.22 MB
conv bs:  1 chans: 12 ->  32 k:3              0.55 ms (  103.44 GFLOPS     2.66 GB/s) in torch,    0.87 ms (   65.35 GFLOPS     1.68 GB/s) in tinygrad,    1.58x slower      56.62 MOPS     1.46 MB
partial_sum                     4096x 4096    1.00 ms (   16.82 GFLOPS    67.54 GB/s) in torch,    0.94 ms (   17.96 GFLOPS    72.10 GB/s) in tinygrad,    0.94x faster      16.84 MOPS    67.63 MB
permute                         1024x 1024    0.47 ms (    2.25 GFLOPS    18.00 GB/s) in torch,    0.46 ms (    2.27 GFLOPS    18.18 GB/s) in tinygrad,    0.99x faster       1.05 MOPS     8.39 MB
permute                         4096x 4096    4.46 ms (    3.76 GFLOPS    30.08 GB/s) in torch,    1.20 ms (   13.93 GFLOPS   111.44 GB/s) in tinygrad,    0.27x faster      16.78 MOPS   134.22 MB
pow                             2048x 2048    0.88 ms (  261.56 GFLOPS   114.14 GB/s) in torch,    1.12 ms (  206.73 GFLOPS    90.21 GB/s) in tinygrad,    1.27x slower     230.69 MOPS   100.66 MB
relu                            4096x 4096    1.41 ms (   11.92 GFLOPS    95.38 GB/s) in torch,    1.19 ms (   14.13 GFLOPS   113.01 GB/s) in tinygrad,    0.84x faster      16.78 MOPS   134.22 MB
sub                             4096x 4096    1.68 ms (   10.01 GFLOPS   120.17 GB/s) in torch,    1.57 ms (   10.69 GFLOPS   128.30 GB/s) in tinygrad,    0.94x faster      16.78 MOPS   201.33 MB
sum                             2048x 2048    0.67 ms (    6.25 GFLOPS    25.10 GB/s) in torch,    0.58 ms (    7.30 GFLOPS    29.31 GB/s) in tinygrad,    0.86x faster       4.21 MOPS    16.91 MB
sum                             4096x 4096    0.99 ms (   16.93 GFLOPS    68.00 GB/s) in torch,    0.88 ms (   19.05 GFLOPS    76.51 GB/s) in tinygrad,    0.89x faster      16.84 MOPS    67.63 MB
```

As for `test_real_world.py` my m1pro might run out of memory for llama and sd, maybe better to retest them.
```
New:
METAL=1 python3 models/test_real_world.py
test_llama: used 13.49 GB and 486 kernels in 1601.10 ms
.test_sd: used 14.06 GB and 1016 kernels in 2290.58 ms
.train_cifar: used 7.85 GB and 274 kernels in 1812.34 ms

Old:
METAL=1 python3 models/test_real_world.py
test_llama: used 13.49 GB and 486 kernels in 1633.04 ms
.test_sd: used 14.05 GB and 924 kernels in 2424.88 ms
.train_cifar: used 7.55 GB and 236 kernels in 1841.71 ms
```

**Comment by stevenandersonz** [comment]
> As for `test_real_world.py` my m1pro might run out of memory for llama and sd, maybe better to retest them.
If you run it with CI=1 it will test the tiny llama

**Comment by nimlgen** [comment]
> > As for `test_real_world.py` my m1pro might run out of memory for llama and sd, maybe better to retest them.
> 
> If you run it with CI=1 it will test the tiny llama

Thanks. But for llama count of kernels is the same, so the speed should be the same. Llama has reductions with global dims like (242, 64), so it's better not to apply the optimizations, since it just adds more to global/local dims (this amount of threads simply overload gpus, for some gpus performance is unchanged, for some it is worse).

I tested a smaller version of sd (with Tensor(1,4,64,64)->Tensor(1,4,48,48). Overall it's a bit better than original verison:
```
OLD
test_llama: used 13.49 GB and 486 kernels in 1563.78 ms
.test_sd: used 7.30 GB and 924 kernels in 499.63 ms
.train_cifar: used 7.55 GB and 236 kernels in 1830.16 ms

test_llama: used 13.49 GB and 486 kernels in 1617.32 ms
.test_sd: used 7.30 GB and 924 kernels in 501.72 ms
.train_cifar: used 7.55 GB and 236 kernels in 1831.49 ms
  
test_llama: used 13.49 GB and 486 kernels in 1595.49 ms
.test_sd: used 7.30 GB and 924 kernels in 503.38 ms
.train_cifar: used 7.55 GB and 236 kernels in 1826.27 ms

NEW
test_llama: used 13.49 GB and 486 kernels in 1545.04 ms
.test_sd: used 7.30 GB and 928 kernels in 497.91 ms
.train_cifar: used 7.85 GB and 274 kernels in 1786.99 ms

test_llama: used 13.49 GB and 486 kernels in 1537.53 ms
.test_sd: used 7.30 GB and 928 kernels in 498.68 ms
.train_cifar: used 7.85 GB and 274 kernels in 1812.01 ms

test_llama: used 13.49 GB and 486 kernels in 1599.92 ms
.test_sd: used 7.30 GB and 928 kernels in 496.24 ms
.train_cifar: used 7.85 GB and 274 kernels in 1781.60 ms
```

**Comment by g1y5x3** [comment]
> .train_cifar: used 7.85 GB and 274 kernels in 1781.60 ms

`test_real_world` should have reported a failure if the generated kernels is greater than [`236`](https://github.com/tinygrad/tinygrad/blob/2d10e0340e288f9a580b91100ab91afa2fbcfb6e/test/models/test_real_world.py#L90C3-L90C3), not sure why it is passing in CI

File: tinygrad/ops.py [hunk 1]
```diff
@@ -152,8 +152,8 @@ def __call__(self, rawbufs:List[RawBuffer], jit=False, force_wait=False) -> Opti
     return et
 
 class Compiled:
-  def __init__(self, buffer: Type[RawBuffer], codegen, runtime, synchronize=lambda: None):
-    self.buffer, self.codegen, self.runtime, self.synchronize = buffer, codegen, runtime, synchronize
+  def __init__(self, buffer: Type[RawBuffer], codegen, runtime, synchronize=lambda: None, device_info = DeviceInfo()):
```
Comment (dc-dc-dc) [inline tinygrad/ops.py:155]: why not just use None as the default? should simplify the value extraction in `reduce_op` as you can check if device_info is set or not

**Comment by g1y5x3** [comment]
image
Just did a quick test on cifar using `CL` backend on `3090`, working fine.

**Comment by nimlgen** [comment]
> > .train_cifar: used 7.85 GB and 274 kernels in 1781.60 ms
> 
> The kernel counts were introduced due to when #1187 was first merged, it halved computation speed for OpenCL backend on nvidia and amd GPUs. I assume that count should be device agnostic unless you change the way kernels are generated.

It should have the same speed on CUDA/CL as #1187. The difference is that this MR doesn't generate more threads than gpu can handle. If a gpu is loaded enough with threads, no need to use this 2 kernels approach. But since amount of parallel threads depends on the size of gpu (it's about 1024*sm_count), this optimisation is also depends on the gpu it's running.

File: tinygrad/ops.py (hunk 1 above)
Comment (nimlgen) [reply tinygrad/ops.py]: Good catch, thanks

File: tinygrad/lazy.py [hunk 2]
```diff
@@ -209,11 +209,32 @@ def shuffle_and_prune_movement_ops(self, st: ShapeTracker, op: MovementOps, arg:
...
+      reduced_dimensions = [(i, divisor := math.gcd(512, old), divisor/(stride or math.inf)) for i, (old, new, stride) in enumerate(zip(self.shape, new_shape, self.st.real_strides())) if old != new] # type: ignore
+      other_dimensions = [old for (old, new) in zip(self.shape, new_shape) if old == new]
+
+      # heuristic -> choose largest divisor to split on, penalize large strides
+      dimension_to_split, divisor, heuristic = max(reduced_dimensions, key=lambda v: v[2])
+      if heuristic < 1: return self._reduce_op(op, new_shape)
+
+      # When global/local dims is filled well, no need to split making it overloaded.
+      sm_count = 32
+      if Device[self.device].device_info: sm_count = unwrap_optional(Device[self.device].device_info.cores_count_executing_in_parallel, sm_count) # type: ignore
```
Comment (geohot) [inline tinygrad/lazy.py:228]: Can a generic estimate / static value be used for this? Then `DeviceInfo` doesn't have to be in this PR.

**Comment by geohot** [comment]
Are you very confident in this? Last time we merged this, we had to revert.

**Comment by nimlgen** [comment]
No, no. Found what's wrong with SD. Need to fix it

**Comment by geohot** [comment]
I think this is gonna need a lot more testing before we can merge this.

**Comment by nimlgen** [comment]
Tested this PR on some models from examples/. On what I have tested (see full results for details) this PR is never worse. SD is now even a bit faster after #1429 + this PR.

<details>
<summary>Full results</summary>

```
stable_diffusion.py NEW: avg:  1075.24 GFLOPS    20.06 GB/s           total:  1193 kernels  4221.37 GOPS    78.76 GB  3925.99 ms
stable_diffusion.py OLD: avg:  1085.77 GFLOPS    18.94 GB/s           total:  1076 kernels  4372.68 GOPS    76.29 GB  4027.27 ms

train_resnet.py NEW: 8.04it/s
train_resnet.py OLD: 7.71it/s

efficientnet.py: same
hlb_cifar10.py: same
train_efficientnet.py: same
serious_mnist.py: same
transformer.py: same

test_real_world.py:
  OLD:
  test_sd: used 7.30 GB and 924 kernels in 499.92 ms
  train_cifar (smaller 48x48): used 7.55 GB and 236 kernels in 1834.85 ms
  NEW:
  test_sd: used 7.30 GB and 932 kernels in 495.85 ms
  train_cifar (smaller 48x48): used 7.85 GB and 274 kernels in 1807.73 ms
  test_llama: same
```

</details>

**Comment by geohot** [comment]
Hoping you got it with that bug fix. Will merge, if there's no revert in a day or two, $400 bounty is yours! george@tinygrad.org to claim

**Pull request merged**